File: Codeforces/126B Password/B_Password-Hashing.py
#  Codeforces 126B Password
# https://codeforces.com/problemset/problem/126/B

import os, sys, io
sys.stdin = io.StringIO(os.read(0, os.fstat(0).st_size).decode())
input = sys.stdin.readline
DOMAIN = 26
PRIME = 2**31-1

s = input().rstrip()
n = len(s)

pows = [0]*n
pows[0] = 1

for i in range(1, len(pows)):
    pows[i] = (pows[i-1]*DOMAIN) % PRIME

hashes = [0] * (n+1)

for i in range(n):
    hashes[i+1] = (hashes[i] + (ord(s[i]) - 95) * pows[i]) % PRIME

boundaries = []

for i in range(1, n):
    suffix_hash = ( hashes[n] - hashes[n-i] + PRIME ) % PRIME
    prefix_hash = (hashes[i] * pows[n-i]) % PRIME
    if suffix_hash == prefix_hash:
        boundaries.append((i, suffix_hash))
        
def find(k, target_hash):
    for i in range(1, len(hashes)-k-1):
        hash = (hashes[i+k] - hashes[i] + PRIME) % PRIME
        hash = (hash * pows[n-(i+k)]) % PRIME
        if hash == target_hash:
            return True

    return False

low = 0
high = len(boundaries) - 1
longest_found = 0

while low <= high:
    mid = low + (high-low) // 2
    len_boundary, hash_boundary = boundaries[mid]
    if find(len_boundary, hash_boundary):
        longest_found = len_boundary
        low = mid + 1
    else:
        high = mid - 1
        

if longest_found == 0:
    print("Just a legend")
    exit(0)
    
print(s[:longest_found])

File: Codeforces/126B Password/B_Password-Hashing.cpp
//  Codeforces 126B Password
// https://codeforces.com/problemset/problem/126/B

#include "bits/stdc++.h"
using namespace std;

const int DOM = 26;
const int PRIME = INT_MAX;

int main() {
    ios::sync_with_stdio(0);
    cin.tie(0);

    string s; cin >> s;
    int n = s.size();

    vector<long long> pows(n), hashes(n+1);

    pows[0] = 1;
    for(int i = 1; i < n; ++i) {
        pows[i] = (pows[i-1] * DOM) % PRIME;
    }

    for ( int i = 0; i < hashes.size()-1; ++i) {
        hashes[i+1] = ( hashes[i] + (s[i]-95)*pows[i] ) % PRIME;
    }

    vector<pair<int, long long>> boundaries;

    for (int i = 1; i < n; ++i) {
        long long suffix_hash = (hashes[n] - hashes[n-i] + PRIME) % PRIME;
        long long prefix_hash = (hashes[i]*pows[n-i] ) % PRIME;
        if (suffix_hash == prefix_hash) {
            boundaries.push_back({i, suffix_hash});
        }
    }

    function<bool(int, long long)> find = [&] (int k, long long target_hash) {
        for(int i = 1; i+k < hashes.size()-1; ++i) {
            long long hash = (hashes[i+k] - hashes[i] +PRIME) % PRIME;
            hash = (hash*pows[n-(i+k)]) % PRIME;
            if (hash == target_hash) {
                return true;
            }
        }
        return false;
    };

    int low = 0;
    int longest_found = 0;
    int high = boundaries.size()-1;
    while (low <= high) {
        int mid = low + (high - low) / 2;
        int k = boundaries[mid].first;
        int target_hash = boundaries[mid].second;
        if (find(k , target_hash)) {
            longest_found = k;
            low = mid + 1;
        } else {
            high = mid - 1;
        }
    }


    if (longest_found == 0) {
        cout << "Just a legend\n";
        return 0;
    }

    cout << s.substr(0, longest_found) << endl;


}
